Api: Track partition statistics via TableMetadata

ajantha-bhat opened the pull request:
- Introduce `PartitionStatisticsFile` as per the Spec.
- Tracking `PartitionStatisticsFile` in a same way as how `StatisticsFile` is already tracked.

Fixes #8457

ajantha-bhat commented on api/src/main/java/org/apache/iceberg/PartitionStatisticsFile.java:
```diff
@@ -0,0 +1,41 @@
+ * <p>Statistics are informational. A reader can choose to ignore statistics information. Statistics
+ * support is not required to read the table correctly.
+ */
+public interface PartitionStatisticsFile extends Serializable {
```
As per the spec

https://github.com/apache/iceberg/blob/main/format/spec.md#partition-statistics

aokolnychyi commented:
Ack, will review this week.

ajantha-bhat commented:
> Ack, will review this week.

Thanks. 
Also, please include https://github.com/apache/iceberg/pull/9170 in the review

aokolnychyi commented:
I started looking on Friday but got distracted. I will try to finish by end of Wed.

aokolnychyi reviewed (commented):
I did one pass. Overall, this seems solid. I'll need to do another pass with fresh eyes.

Are saving the rest of changes in engines to keep these files during cleanup for future PRs? Shall we test the core expiry of snapshots in this PR? It would be unfortunate to add partition stats files and let the expiry process remove them.

ajantha-bhat commented:
@aokolnychyi: Thanks for the review. 

- I see that most of the questions or comments related to why puffin stats followed that style. We can fix in this PR for partition stats and later back port to puffin too.  
- I didn't add expire snapshots and remove orphan files test code (but have test case of RemoveSnapshots API) as I was thinking to keep the scope of PR to non-Spark. I will have a PR that depends on this PR which tests these function. So, this PR can be merged. 

Got a little busy week. I will finish addressing comments and the follow up Spark PR for expire snapshots and remove orphan files by Monday. 

Meanwhile you can also review the independent PR (Util for partition stats reading and writing) : https://github.com/apache/iceberg/pull/9170

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:
```diff
@@ -137,7 +137,9 @@ public static List<String> manifestListLocations(Table table, Set<Long> snapshot
    * @return the location of statistics files
+   * @deprecated use the {@code allStatisticsFilesLocations(table)} instead.
    */
+  @Deprecated
```
removed the changes in this file as the new APIs added was not called from anywhere (this is needed for spark action). 

I will raise a follow up PR for Spark actions to consider partition stats file for remove orphan files and expire snapshots action.

ajantha-bhat replied:
Since, it was needed only for spark action, moved these changes into a separate PR with the tests for expire snapshots and remove orphan files. 

https://github.com/apache/iceberg/pull/9284

ajantha-bhat commented on .palantir/revapi.yml:
```diff
@@ -873,6 +873,10 @@ acceptedBreaks:
       justification: "Static utility class - should not have public constructor"
   "1.4.0":
+    org.apache.iceberg:iceberg-api:
+    - code: "java.method.addedToInterface"
```
added as default

ajantha-bhat commented:
@aokolnychyi: Fixed all the comments and also opened a new Spark module PR (which is dependent on this) to ensure partition stats are considered for GC (expire snapshots and remove orphan files). I will rebase that PR once this is merged.

https://github.com/apache/iceberg/pull/9284

aokolnychyi commented on core/src/main/java/org/apache/iceberg/TableMetadataParser.java:
```diff
@@ -561,4 +576,21 @@ private static List<StatisticsFile> statisticsFilesFromJson(JsonNode statisticsF
   }
+
+  private static List<PartitionStatisticsFile> partitionStatisticsFilesFromJson(
+      JsonNode partitionStatisticsFilesList) {
```
I don't mind using longer variables but I wonder whether a shorter version and staying on one line in some places would be more readable in this method.

aokolnychyi reviewed (commented):
This looks close. I did another pass, I'll need to check the tests with fresh eyes.

aokolnychyi commented:
@ajantha-bhat, I'll take a look at other PRs once this is in. I feel this one is almost ready to go.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/TableMetadata.java:
```diff
@@ -946,15 +955,25 @@ private Builder(TableMetadata base) {
       this.previousFileLocation = base.metadataFileLocation;
       this.previousFiles = base.previousFiles;
       this.refs = Maps.newHashMap(base.refs);
-      this.statisticsFiles =
```
Currently one stats file per snapshot. But I think in future it may track multiple files too. 

I followed the same pattern as existing puffin files. I want to keep the interfaces consistent.

ajantha-bhat commented on api/src/main/java/org/apache/iceberg/PartitionStatisticsFile.java:
```diff
@@ -0,0 +1,38 @@
+ * <p>Statistics are informational. A reader can choose to ignore statistics information. Statistics
+ * support is not required to read the table correctly.
+ */
+public interface PartitionStatisticsFile extends Serializable {
```
Ack. We can add it back if required during an end to end implementation.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/TableMetadata.java:
```diff
@@ -946,15 +955,25 @@ private Builder(TableMetadata base) {
       this.sortOrdersById = Maps.newHashMap(base.sortOrdersById);
     }
 
+    private static Map<Long, List<StatisticsFile>> statsFileBySnapshotID(TableMetadata base) {
```
moved

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/TableMetadata.java:
```diff
@@ -946,15 +955,25 @@ private Builder(TableMetadata base) {
+          .collect(Collectors.groupingBy(StatisticsFile::snapshotId));
+    }
+
+    private static Map<Long, List<PartitionStatisticsFile>> partitionStatsFileBySnapshotID(
```
updated

ajantha-bhat commented:
@aokolnychyi: I have handled all the new comments. Thanks again for the review.

ajantha-bhat commented:
Re-triggering build due to flaky test in Flink

https://github.com/apache/iceberg/actions/runs/7222400549/job/19679307743?pr=8502

aokolnychyi commented on core/src/main/java/org/apache/iceberg/SetPartitionStatistics.java:
```diff
@@ -0,0 +1,68 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+
+public class SetPartitionStatistics implements UpdatePartitionStatistics {
```
Does this have to be public?

aokolnychyi reviewed (approved):
This looks good to me. I had only one comment about the access modifier for the base implementation class.

ajantha-bhat commented:
@aokolnychyi: Thanks for the detailed review and merge.

The next small PR to review will be https://github.com/apache/iceberg/pull/9284 (Spark 3.5: Ensure that partition stats files are considered for GC procedures)

>  I had only one comment about the access modifier for the base implementation class.

I will check this in a followup. Again this follows the style from puffin files. 
I will also handle the applicable comments from this PR to stats files code (puffin files) in a follow up PR.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/TableMetadataParser.java:
```diff
@@ -481,6 +488,13 @@ public static TableMetadata fromJson(String metadataLocation, JsonNode node) {
       statisticsFiles = ImmutableList.of();
     }
 
+    List<PartitionStatisticsFile> partitionStatisticsFiles;
+    if (node.has(PARTITION_STATISTICS)) {
+      partitionStatisticsFiles = partitionStatsFilesFromJson(node.get(PARTITION_STATISTICS));
+    } else {
+      partitionStatisticsFiles = ImmutableList.of();
+    }
+
```
The way partition stats tracked and added to table metadata is same as puffin files right now. 

The stats are optional, so even if we lose it. Planner won't return wrong query results. 

However these stats can be helpful to improve query performance. We are planning to provide a call procedure, compute partition stats. Which will check the last snapshot that had partition stats and incrementally compute the stats for remaining snapshots.
